### PR TITLE
Add an all-in-one histogram intrinsic, along with lowering for AArch64

### PR DESCRIPTION
**Pull request by huntergr-arm**
Based on discussion from https://discourse.llvm.org/t/rfc-vectorization-support-for-histogram-count-operations/74788

Current interface is:

llvm.experimental.histogram(<vecty> ptrs, <intty> inc_amount, <vecty> mask)

The integer type used by 'inc_amount' needs to match the type of the buckets
in memory.

The intrinsic covers the following operations:
  * Gather load
  * histogram on the elements of 'ptrs'
  * multiply the histogram results by 'inc_amount'
  * add the result of the multiply to the values loaded by the gather
  * scatter store the results of the add

At this stage, we'd mostly like to discuss which type of intrinsic would be preferable (this one, or the original proposed in the RFC). There will be work on the LoopVectorize side to recognize histogram operations proceeding independently of this patch, at least for now.

### CODE REVIEW

**Comment by RonDahan101** [comment]
This is a great improvement! I would be glad if you could consider supporting these two cases as well:
* Histograms with saturated values.
* More operations instead of only 'add' (Min/Max/Set).

**Review by SamTebbs33 — APPROVED** [review]
This looks good to me as is and I support the all-in-one approach. I think it would be worth addressing the FIXMEs that are simple enough to do in this patch (moving out code to their own functions and the operand assertion in `getMaskedHistogram`).

File: llvm/include/llvm/IR/Intrinsics.td [hunk 1]
```diff
@@ -1848,6 +1848,13 @@ def int_experimental_vp_strided_load  : DefaultAttrsIntrinsic<[llvm_anyvector_ty
                                llvm_i32_ty],
                              [ NoCapture<ArgIndex<0>>, IntrNoSync, IntrReadMem, IntrWillReturn, IntrArgMemOnly ]>;
 
+// Experimental histogram
+def int_experimental_histogram : DefaultAttrsIntrinsic<[],
+                             [ llvm_anyvector_ty, // Vector of pointers
+                               llvm_anyint_ty,    // Increment
+                               LLVMScalarOrSameVectorWidth<0, llvm_i1_ty>], // Mask
+                             [ IntrNoSync, IntrWillReturn ]>;
```
Comment (arsenm) [inline llvm/include/llvm/IR/Intrinsics.td:1856]: IntrNoSync and IntrWillReturn are implied by DefaultAttrsIntrinsic

**Review by paschalis-mpeis — APPROVED** [review]
Looks good. Agree with proceeding with some of the FIXMEs.

**Comment by huntergr-arm** [comment]
Rebased, addressed the easy FIXMEs.

I made the intrinsic form extensible, so that you specify the update operation instead of just assuming it's an addition. Have added to the LangRef.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -6673,6 +6677,56 @@ SDValue AArch64TargetLowering::LowerOperation(SDValue Op,
...
+    SDValue IncSplat = DAG.getSplatVector(MemVT, DL, Inc);
+    SDValue Ops[] = {Chain, PassThru, Mask, Ptr, Index, Scale};
+
+    // Set the MMO to load only, rather than load|store.
+    MachineMemOperand *GMMO = HG->getMemOperand();
+    GMMO->setFlags(MachineMemOperand::MOLoad);
+    ISD::MemIndexType IndexType = HG->getIndexType();
+    SDValue Gather =
+                DAG.getMaskedGather(DAG.getVTList(MemVT, MVT::Other), MemVT, DL,
+                                    Ops, HG->getMemOperand(),
```
[the remark this answers is not in the document]
Comment (huntergr-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: MemSDNodes have a member variable of a pointer to a MachineMemOperand, so to correctly have separate MOLoad semantics for the gather and MOStore semantics for the scatter we need two different MMO instances.

(I'm not sure if it matters if we just let both be `MOLoad | MOStore` at this point, I'm just being a bit cautious)

File: llvm/docs/LangRef.rst [hunk 3]
```diff
@@ -19030,6 +19030,65 @@ will be on any later loop iteration.
...
+   1. Perform a cross-vector histogram operation on the ``ptrs`` operand,
+      or a set of index values if it can be decomposed into a base pointer
+      with smaller indices matching the type of ``inc``.
+   2. Multiply the result by the ``inc`` operand.
+   3. Add the result to the values loaded from memory
+3. Scatter the result of the update operation to the memory locations from
+   the ``ptrs`` operand.
+
+The ``mask`` operand will apply to at least the gather and scatter operations,
+and potentially the update if supported.
```
Comment (jayfoad) [inline llvm/docs/LangRef.rst:19090]: "potentially the update if supported" sounds like an implementation optimization. I have a slight preference to keep the semantics section as simple as possible by omitting it. Same for the "or a set of index value..." part above.

File: llvm/docs/LangRef.rst [hunk 4]
```diff
@@ -19030,6 +19030,65 @@ will be on any later loop iteration.
...
+
+Arguments:
+""""""""""
+
+The first argument is a vector of pointers to the memory locations to be
+updated. The second argument is a scalar used to update the value from
+memory; it must match the type of value to be updated. The final argument
+is a mask value which will exclude that vector element from being updated,
+and will exclude it from any cross-lane calculations to determine the
+final values for each memory location.
```
Comment (jayfoad) [inline llvm/docs/LangRef.rst:19067]: I'm not sure that the "and will exclude it from ..." part really adds anything useful to the description?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (huntergr-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Ah, I think I see what you meant now; the diff wasn't quite clear after the forced push. Done.

**Comment by huntergr-arm** [comment]
Rebased, addressed comments. I've added scalarization of the intrinsic for targets which don't have an appropriate vector instruction to map this to.

File: llvm/docs/LangRef.rst [hunk 5]
```diff
@@ -19068,6 +19068,60 @@ will be on any later loop iteration.
 This intrinsic will only return 0 if the input count is also 0. A non-zero input
 count will produce a non-zero result.
 
+'``llvm.experimental.vector.histogram.*``' Intrinsics
```
Comment (mgabka) [inline llvm/docs/LangRef.rst:19071]: nit: looks like a singular form Intrinsics is used every where else.

**Comment by RKSimon** [comment]
@huntergr-arm This is failing on EXPENSIVE_CHECKS builds - please can you take a look?
https://lab.llvm.org/buildbot/#/builders/16/builds/65450

**Comment by huntergr-arm** [comment]
Hi @RKSimon -- I already have a patch for that: https://github.com/llvm/llvm-project/pull/92095
I should have mentioned it here though.

**Pull request merged**